src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.NativeLayout.cs

```diff
@@ -212,6 +212,18 @@ private void CreateNodeCaches()
             // of just necessary. (Which is what the actual templates signatures will ensure)
             public IEnumerable<IDependencyNode> TemplateConstructableTypes(TypeDesc type)
             {
+                // Array types are the only parameterized types that have templates
+                if (type.IsSzArray && !type.IsArrayTypeWithoutGenericInterfaces())
+                {
+                    TypeDesc arrayCanonicalType = type.ConvertToCanonForm(CanonicalFormKind.Specific);
+
+                    // Add a dependency on the template for this type, if the canonical type should be generated into this binary.
+                    if (arrayCanonicalType.IsCanonicalSubtype(CanonicalFormKind.Any) && !_factory.NecessaryTypeSymbol(arrayCanonicalType).RepresentsIndirectionCell)
+                    {
+                        yield return _factory.NativeLayout.TemplateTypeLayout(arrayCanonicalType);
+                    }
+                }
+
                 while (type.IsParameterizedType)
```

Comment:
This while loop looks a bit suspect. Let's say we have `T[][][]`. Do we need to ensure the templates for all `T[]`, `T[][]` and `T[][][]` ?

Reply:
Yeah, maybe we're at risk of missing the `__Canon[]` template here, but that will probably never happen in practice.

Template for `T[]` is `X[]` (where X is the thing T canonicalizes into).

Template for T[][][] is `__Canon[]`, no matter the T (because the element type is a reference type).

Reply:
(This whole TemplateConstructableTypes business is probably just someone getting tired of fixing small bugs within the dependencies. I think it creates a lot more templates than really needed. I would like to get rid of it at some point.)